src/protocols/secure_channel/CASESession.cpp

```diff
@@ -1516,6 +1546,13 @@ CHIP_ERROR CASESession::SendSigma3c(SendSigma3Data & data, CHIP_ERROR status)
     return err;
 }
 
+void CASESession::OnSigma3_AsyncFailure(CHIP_ERROR status)
+{
+    SendStatusReport(mExchangeCtxt, kProtocolCodeInvalidParam);
```

Comment:
This is not OK, and will fail threasafety asserts.  You _have_ to deal with exchanges on the Matter thread, or at least while holding the locks.... and note that not all platforms allow you to have locks.

This part needs some careful thought.

What we probably need is some guarantee that a platform allows _at least one_ of locking the Matter stack and infallible dispatch to the Matter queue, auditing of our platforms to that effect, and documentation in platform manager to that effect.  And then here, since we know the dispatch failed (and we would need to restrict to that case!), we can take the relevant locks, etc.

Reply:
Sorry for the confusion. I mistakenly assumed that `ScheduleWork` and `ScheduleBackgroundWork` are basically same, but it appears that they operate differently and may not work here. I'll check what can be done. Thank you for bringing this to my attention.